Smartglass

SonixApache opened the pull request:
![smartglass](https://user-images.githubusercontent.com/8468269/30299697-27c4106e-9751-11e7-981d-900f36b453ac.gif)
General idea:
https://streamable.com/rk43v
Applying to windows:
https://streamable.com/x7uit
Button functionality:
https://streamable.com/qxj3x
Windoors:
https://streamable.com/u6td8

### Functionality:
- Smartglass can be ~~powered~~ toggled to be transparent, else is opaque
- Add wired glass (glass + wire) to any window or windoor to convert to smartglass
- Use multitool menu to get/set ID tag for buttons - the command is `toggle_transparency`
- Use multitool menu to toggle transparent/opaque
- Unanchor a window to remove the wired glass from it

🆑 
 - rscadd: Added smartglass; glass that can be digitally switched between transparent and opaque. Apply a wired glass tile to a window or windoor to create it, then multitool it for setup.

Duny- commented:
Button functionality WHEN

SonixApache commented:
Gonna work on it more tomorrow or so, whenever my friends don't want to play videogames anymore.
Weird how it fails on travis even though it compiled locally, though.

SonixApache commented on vgstation13.dme:
```diff
@@ -2222,7 +2223,7 @@
 #include "maprendering\maprendering.dm"
 #include "maps\_map.dm"
-#include "maps\tgstation.dm"
+#include "maps\test_tiny.dm"
```
whoops

Ratvar commented:
add smartwalls

SonixApache commented:
you have to go back

Unknown-as-Captain commented:
Cool

Intigracy commented on code/game/machinery/smartglass.dm:
```diff
@@ -0,0 +1,38 @@
+
+/obj/machinery/smartglass_electronics
+	name = "internal smartglass electronics"
+	desc = "This inside a pane of smart glass. How are you seeing this?"
```
This missing a word.

Intigracy commented on code/game/objects/structures/window.dm:
```diff
@@ -309,10 +318,29 @@ var/list/one_way_windows
 		one_way_windows.Add(src)
 		P.use(1)
-		to_chat(user, "<span class='notice'>You place a sheet of plastic over the window.</span>")
+		to_chat(user, "<span class='notice'>You place a sheet of plastic over the [src].</span>")
```
\the [src] or you're gonna get the the window

Exxion replied:
I still have a hard time understanding how they could have possibly thought that was a good idea
"Hey, let's add text macros to add the to improper nouns, etc.!"
"Better idea: We do that, but also make them present by default, thus making adding words to your shit an opt-out procedure, as well as making several of those macros completely superfluous!"

Intigracy commented on code/game/objects/structures/window.dm:
```diff
@@ -309,10 +318,29 @@ var/list/one_way_windows
+			to_chat(user, "<span class='notice'>This [src] already has electronics in it.</span>")
+			return
+		LT.use(1)
+		to_chat(user, "<span class='notice'>You add some electronics to the [src].</span>")	
```
the the window

Intigracy commented on code/game/objects/structures/window.dm:
```diff
@@ -309,10 +318,29 @@ var/list/one_way_windows
+	if(istype(W, /obj/item/stack/light_w))
+		var/obj/item/stack/light_w/LT = W
+		if (smartwindow)
+			to_chat(user, "<span class='notice'>This [src] already has electronics in it.</span>")
```
This the window

SonixApache replied:
There's no `\this`,  right?

Intigracy replied:
\the, \The // insert 'the' (if needed)
\a, \an,   // insert 'a(n)' or 'some' (if needed)
\A, \An

http://www.byond.com/docs/ref/info.html#/DM/text/macros

SonixApache commented:
- Fixed grammar
- Added a toggle proc for the one-way windows
- Added a machine inside the glass that handles the multitool menu
- Menu options don't actually work because 1: wrong procs and 2: z-level issues
- Added a chat notice when you're out of range when using multitools

SonixApache commented:
- made radio functionality work
- some other stuff....?

SonixApache commented:
- Fixed the one-way toggle button on the multitool menu
- Limited the access range to one tile, like most machinery

Only issue I have now is that it doesn't actually draw power from APCs, and thus losing power doesn't toggle the windows back into opaque mode.

SonixApache commented:
Added functionality to windoors.
https://puu.sh/xvEqp/8d8107db53.mp4

SonixApache commented:
Added functionality to windoors.
https://puu.sh/xvEqp/8d8107db53.mp4
Can also remove smartglass functionality by dislocating a window.

SonixApache commented:
Okay, two things to do.
- [ ] Datumize the internal electronics to use `power_connection` to solve power issues
- [ ] Fix multitool cloning

CptWad commented:
ho ho holy shit
I have excellent ideas for this

SonixApache commented:
Running into a lotta issues with making the internal machinery operate on power. Might just drop those features if I can't figure it out this week.

Exxion commented:
I believe all you have to do is manually update the contained machinery's `areaMaster`

SonixApache commented:
Right now the power functionality is broken regardless. Need to figure that part out first.
Tried manually adding it to the `areaMaster` before, it resulted in the machinery being teleported outta the glass and into a random spot in the room.

Exxion commented:
No, change the machinery's `areaMaster` var

```
/obj/machinery/smartglass_electronics/use_power()
    areaMaster = GLASS.areaMaster
    return ..()
```
Except replace the spaces with tabs

SonixApache commented:
What the hell? I thought `use_power` was a var, not a proc?

Exxion commented:
It's both

Exxion commented:
Alright, we don't really have any var capitalization conventions
But all-caps is like the one format we don't have
All-caps is generally used only for defines

Intigracy commented:
I wish we had standardization too but what can ya do

N3X15 commented on code/game/machinery/doors/windowdoor.dm:
```diff
@@ -20,12 +20,14 @@
-		src.icon_state = "[src.icon_state]"
-		src.base_state = src.icon_state
+	if ((istype(req_access) && req_access.len) || istext(req_access))
+		icon_state = "[icon_state]"
```
what

N3X15 commented on code/game/machinery/doors/windowdoor.dm:
```diff
@@ -100,50 +113,53 @@
-	flick(text("[]opening", src.base_state), src)
+	if(!operating) //in case of emag
+		operating = 1
+	flick(text("[]opening", base_state), src)
```
Dear god how old is this code if it still has text()?!

N3X15 commented on code/game/machinery/smartglass.dm:
```diff
@@ -0,0 +1,131 @@
+	
+	//Smartglass vars
+	var/smart_transparency = 0
+	var/obj/structure/window/GLASS //Ref to the window we're in
```
Don't use ALLCAPS variable names, that's for constants.

SonixApache replied:
roger

SonixApache commented:
Varname changed to `Ourwindow`.
Reverted all the `src` removal to make the diffs more readable.
